backend/jottacloud/jottacloud.go

```diff
@@ -748,6 +759,59 @@ func (o *Object) Open(options ...fs.OpenOption) (in io.ReadCloser, err error) {
 //
 // The new object may have been created if an error is returned
 func (o *Object) Update(in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (err error) {
+	md5String, err := src.Hash(hash.MD5)
+	if err != nil || md5String == "" {
+		// if the source can't provide a MD5 hash we are screwed and need to calculate it ourselfs
+		// we read the entire file and cache it in the localdir and send it AFTERwards
+
+		// we need a MD5
+		md5Hasher := md5.New()
+		// use the teeReader to write to the local file AND caclulate the MD5 while doing so
+		teeReader := io.TeeReader(in, md5Hasher)
+
+		// don't cache small files on disk to reduce wear of the disk
+		if src.Size() > int64(o.fs.opt.MD5MemoryThreshold) {
+			// create the cache file
+			tempFile, err := ioutil.TempFile("", "rclone-jcmd5-")
+			if err != nil {
+				return err
+			}
+
+			// reade the ENTIRE file to disc and calculate the MD5 in the process
+			if _, err = io.Copy(tempFile, teeReader); err != nil {
+				return err
+			}
+			// jump to the start of the local file so we can pass it along
+			if _, err = tempFile.Seek(0, 0); err != nil {
+				return err
+			}
+
+			// clean up the file after we are done downloading
+			defer func() {
```

Comment:
You are absolutely right, fix commited

Reply:
```
cleanup := func() {
  _ = tempFile.Close()
  _ os.Remove(tempFile.Name())
}
defer cleanup()
atexit.Register(cleanup)
```
Wouldn't that just create a memory leak for long running tasks because we collecting and keeping all tempfile objects up until rclone exits.
There should be just one cleanup function the removes the files by using a list of open tempfiles and closing and removing them.

Reply:
Good point. How about avoiding capturing `tempFile` in the cleanup function that is passed to `atexit.Register`? 

```
tempName := tempFile.Name()
atexit.Register(func() { _ = os.Remove(tempName) })
```

The OS will clean up any open file descriptors when the program terminates anyway.

Reply:
In the code I posted, it deletes the file after opening.  On all OSes except Windows this will mean that we use a temporary anonymous file which will get removed when it is closed or the program exits unexpectedly.

https://github.com/ncw/rclone/commit/b510c70c1ef2bd41eee962dc04fbbe0880edebd6#diff-9d62ca00aab4a67defb33a6249776bc3L1263

For the Windows case - the temporary file is in the users TEMP directory which there are friendly Windows way of cleaning.

Using atexit would get some of the additional cases (but not all) that can happen when rclone exits so I'm not sure it is worth it.